### DISCUSSION TITLE
서드파티 API 를 통해 가져온 대용량 데이터 처리시 SELECT 쿼리의 성능 병목 현상 조사와 부하 테스트 및 성능 개선

### DISCUSSION THREAD

**Discussion by grulla99** [100-hours-a-week/discussions#12] [Deepdive]
# 질문
## 제 딥다이브 주제에 대해 생각해봤을 때 아래 내용이 나올 수 있는 딥다이브 방식과 방법밖에 떠오르지 않습니다. 혹시 추가할만한 내용이나 부족한 부분이 있을까요? 검토 부탁드립니다.

# 서드 파티 API를 통해 가져온 대용량 데이터 처리

### 대용량 데이터 사용처

1. 가격 추이 비교에 사용 
    - 일단 MySQL에 저장하는 쿼리 ( Raw 데이터를 저장할 예정 )
    - 저장 했던 데이터 중 날짜, 출발 공항, 도착 공항, 가격을 가지고 가격 추이 그래프 형성
2. 가격 변동 예측에 사용 
    - 1번에서 뽑아온 데이터를 바탕으로 AI 혹은 알고리즘을 통해 가격 변동 예측 프로세스

### 방법

1. 데이터 처리 속도가 느리다면 정형화된 데이터를 분산시켜서 관리하는건 어떤지?
    - 데이터를 통째로 보관해놓는 테이블과 가격추이나 변동예측에 사용하는 데이터를 분리시켜 보관 후 비교
2. Amadeus 에서 받아온 데이터는 데이터 형식이 MySQL로 처리하기에는 복잡한 부분이 있어 NoSQL을 사용해서 데이터 처리 속도 비교해보기

------------------------------------------------------------------------------------------------------------------------

# SELECT 쿼리 수행 동작을 조사하고 부하테스트 및 성능 개선

## 선택 이유 
-  SELECT 쿼리가 데이터베이스에서 어떻게 수행되는지를 깊이 이해하는 것은 프로젝트의 성능 최적화, 복잡한 데이터 구조의 효율적인 관리, 실시간 데이터 처리, 대규모 트래픽 처리에 직접적인 기여를 하며, 이러한 이유로 SELECT 쿼리 수행의 내부 과정을 깊이 있게 분석하는 것이 필수적 생각하여 선택했습니다.
-  ngrinder, jmeter를 사용하여 부하테스트를 진행하고, 성능개선을 해볼 생각입니다.

### 데이터베이스 기본 구조
![image](https://github.com/user-attachments/assets/1e3b69c0-e072-4cd0-8577-8fe8a8baf023)

### MySQL 엔진 구조
![image](https://github.com/user-attachments/assets/0abdaa9d-1f6c-4638-896e-c0ac63bcb5d0)

1. 쿼리 파서
    1. 쿼리 문장을 MySQL이 인식할 수 있는 최소 단위인 토큰으로 분리 후 트리구조를 만듬
    2. 전처리기에게 전달
    3. SQL 문법 오류는 이 과정에서 발견되어 사용자에게 여러 메세지가 전달
2. 전처리기
    1. 전달받은 파서 트리를 기반으로 테이블 이름이 존재하는지, 컬럼 이름이 존재하는지, 접근 권한이 있는지 등을 확인
    2. 이 부분에서 문제가 발견되면 오류처리
3. 옵티마이저
    1. 쿼리를 최적화해서 최종 실행계획 완성
    2. `옵티마이저가 더 나은 선택을 할 수 있게 유도하는 방법` 조사
4. 쿼리 실행기
    1. 스토리지 엔진이 제공하는 핸들러 API를 가지고 옵티마이저가 계획해준 작업 목록을 수행
5. **핸들러 엔진**
    1. MySQL 서버에서 MySQL 엔진은 하나지만 스토리지 엔진은 여러개를 동시에 사용할 수 있다.
    2. MySQL 엔진은 스토리지 엔진이 제공하는 API 를 이용하여 스토리지 엔진에게 CRUD 직업을 위임.
    3. 이 때 스토리지 엔진이 제공하는 api를 핸들러 api 라고 한다. 
    4. 핸들러 규칙을 만족하도록만 구현하면 누구든지 스토리지 엔진을 만들고 MySQL 서버에 추가해서 사용할 수 있다.( **플러그인 모델 )**
        1. 스토리지 엔진 외에도 전문 검색 엔진을 위한 검색어 파서, 사용자 인증 등을 위한 플러그인 등 다양한 기능들이 플러그인으로 구현되어 제공
6. MySQL이 기본적으로 제공하는 스토리지 엔진
    1. InnoDB
```
// user 테이블에 대한 CRUD를 innoDB 스토리지 엔진이 처리하도록 한다.
// 특정 테이블에 특정 스토리지 엔진을 지정가능
create table user(
  id bigint,
  name varchar(50),
  age int
) engine=innodb;
```
<details>
  <summary>MySQL 스토리지 엔진</summary>

- MySQL은 각 데이터베이스(스키마)를 파일시스템 안의 데이터 디렉터리의 하위 디렉터리로 저장한다.
- 테이블을 생성하면 MySQL은 테이블 이름과 같은 이름을 가진 .frm 파일을 만들고 그 안에 테이블 정의 정보를 저장한다.
- flight 이라는 테이블을 만들면 MySQL은 MyTable.frm에 테이블 정의를 저장한다.
- MySQL 은 데이터베이스 이름과 테이블 정의를 저장하는 데 파일시스템을 사용하므로 대/소문자 구분은 플랫폼에 따라 결정된다. ( 윈도우 MySQL 인스턴스에서 테이블과 데이터베이스 이름에는 대/소문자 구분이 없으나 유닉스 계열 시스템에서는 대/소문자를 구분)

### InnoDB  엔진

- 트랜잭션을 처리하기 위해 고안됐다. 대부분 롤백되지 않고 완료(정상 종료) 짧은 트랜잭션이 많은 경우를 처리하기 좋게 되어 있다.
- InnoDB는 가장 많이 사용되는 Sorage Engine으로 ACID 트랜잭션을 지원하는 대표적인 MySQL의 대표적이고 뛰어난 성능과 장애 복구 기능을 가진 엔진이다. 결제 정보와 같이 정보 무결성을 가져야하고 손실되면 안되는 중요한 데이터를 필요로 할 때 사용한다.
- InnoDB 테이블은 클러스터 인덱스위에 구성되어 있다. InnoDB의 인덱스 구조는 대부분의 MySQL 스토리지 엔진의 인덱스 구조와는 매우 상이하다.(innodb가 클러스터드 인덱스를 사용, 하지만 다른 스토리지 엔진은 일반적으로 비클러스터드 인덱스를 사용한다.) 매우 신속한 기본 키 조회가 가능하다.
- 보통 기본 키가 클러스터드 인덱스로 사용된다. 테이블에 Primary Key가 정의되지 않은 경우, InnoDB는 내부적으로 숨겨진 Row ID를 사용하여 클러스터드 인덱스를 생성한다.
- MVCC(Mulit Version Concurrency Control) 기술을 이용해 락을 걸지 않고 읽기 작업을 수행하여 락을 걸지 않기 때문에 InnoDB에서 읽기 작업은 다른 트랜잭션이 가지고 있는 락을 기다리지 않아도 된다.
- MySQL 서버가 시작될 때 완료되지 못한 트랜잭션이나 디스크에 일부만 기록된 트랜잭션 등에 대한 일련의 복구작업이 자동으로 진행.
- 그래프 기반의 데드락 체크 방식을 사용하므로 데드락이 발생함과 동시에 바로 감지되고, 감지된 데드락은 관련 트랜잭션 중 ROLLBACK이 가장 용이한 트랜잭션을 자동적으로 강제 종료한다. 그래서 데드락 때문에 쿼리가 타임아웃 슬로우 쿼리로 기록되는 경우는 많지 않다.
- Clustered Index 와 Non Clustered Index
    
    <aside>
   Clustered Index 와 Non Clustered Index
    
    - 인덱스는 데이터를 빠르게 검색할 수 있게 해주는 객체. 칼럼을 정렬한 후 데이터를 빠르게 찾을 수 있도록 도와준다.
    - 하지만 인덱스를 무작정 생성한다고 좋은 방법은 아니다.
    - 인덱스를 생성하면 인덱스를 위한 디스크 공간이 필요하고, 인덱스를 가진 테이블에 DML 작업을 할 경우 더 많은 비용과 시간이 필요하다.
    - 인덱스 생성 시 해당 테이블의 요도를 정확하게 파악한 후 상황에 맞게 적적한 칼럼으로 Clustered Index 와 Non Clustered Index를 구성해야한다.
    
    1. Clustered Index
    - 테이블에 1개씩만 허용
    - 물리적으로 행을 재배열
    - PK 설정 시 그 칼럼은 자동으로 클러스터드 인덱스가 만들어짐
    - 인덱스 자체의 리프 페이지가 곧 데이터, 즉 테이블 자체가 인덱스. (따로 인덱스 페이지를 만들지 않는다.)
    - 데이터 입력, 수정, 삭제 시 항상 정렬 상태를 유지.
    - Non Clustered Index 보다 검색속도는 더 빠르다. 하지만 데이터의 입력, 수정, 삭제는 느리다.
    - 30% 이내에서 사용해야 좋은 선택도를 가진다.
    
    2. Non Clustered Index
    - 테이블 당 약 240개의 인덱스를 만들 수 있다. 
    - 인덱스 페이지는 로그파일에 저장된다.
    - 레코드의 원본은 정렬되지 않고, 인덱스 페이지만 정렬된다.
    - 인덱스 자체의 리프 페이지는 데이터가 아니라 데이터가 위치하는 포인터(RID)이기 때문에 클러스터형보다 검색속도는 느리지만 데이터 입력, 수정, 삭제는 더 빠르다. 
    - 인댁스를 생성할 대 데이터 페이지는 그냥 둔 상태에서 별도의 인덱스 페이지를 따로 만들기 때문에 용량을 더 차지한다.
    - 3% 이내에서 사용해야 좋은 선택도를 가진다.
    
    ⇒ 클러스터드 인덱스는 데이터 위치를 바로 알기 때문에 그 데이터로 바로 접근할 수 있고, 논 클러스터드 인덱스는 인덱스 페이지를 한 번 거쳐서 데이터에 접근하는 방식이다.
    

    
    </aside>
    

**왜 InnoDB가 클러스터드 인덱스를 사용하는가?**

1.	**ACID 준수와 트랜잭션 지원**:

→ InnoDB는 ACID(Atomicity, Consistency, Isolation, Durability)를 준수하는 트랜잭션을 지원하기 위해 설계됐다. 클러스터드 인덱스는 이러한 특성을 유지하면서도 높은 성능을 제공하는 데 적합하다. 데이터가 물리적으로 정렬되어 있어 특정 키 범위에 대한 빠른 검색이 가능하다.

2.	**외래 키(Foreign Key) 지원**:

→ InnoDB는 외래 키를 지원하는 MySQL의 유일한 스토리지 엔진입니다. 클러스터드 인덱스는 외래 키 제약 조건을 효율적으로 처리할 수 있도록 돕습니다.

3.	**데이터 일관성과 무결성**:

→ 클러스터드 인덱스 구조는 데이터 일관성 유지에 유리합니다. 인덱스와 실제 데이터가 함께 저장되므로, 데이터 손상이나 일관성 문제가 발생할 가능성이 줄어듭니다.

### MYISAM 엔진

- 과거에 MySQL 스토리지 엔진이었다.
- Transaction을 지원하지 않기 때문에 InnoDB보다는 심플하고 기본적으로 빠르지만 동시성 제어가 어렵다는 단점이 있다. 그래서 Read쿼리가 많은 DW환경에서 많이 사용된다.
- 각 테이블을 데이터 파일과 인덱스 파일 두 곳에 주로 저장한다. ( 확장자는 각각 .MYD, .MYL)
- MylSAM 테이블은 동적인 행과 정적인(고정된 길이) 행을 모두 가질 수 있는데 MySQL이 테이블 정의를 토대로 하여 사용할 포맷을 결정한다.
- Key Cache: InnoDB 버퍼풀과 비슷한 역할을 한다. 하지만 이름그대로 인덱스만 대상으로 작동하고 또한 인덱스의 디스크 쓰기 작업에 대해서만 부분적으로 버퍼링한다.
- 행 단위가 아니라 테이블 전체를 잠군다. 읽기 동작은 읽어야할 모든 데이터에 대한 공유된 읽기 권한을 쓰기동작은 배타적 쓰기 잠금 권한을 얻는다. SELECT 쿼리 실행 중에도 이 테이블에 새 행을 삽입 가능하다.( 동시 삽입 )
- 수동 복구
</details>

**Comment by startupcodekr** [comment]
제목에 본인 프로젝트 시나리오에서 문제점과 관련된 내용이 필요합니다!

**Comment by ej31** [comment]
> # 질문
> ## 제 딥다이브 주제에 대해 생각해봤을 때 아래 내용이 나올 수 있는 딥다이브 방식과 방법밖에 떠오르지 않습니다. 혹시 추가할만한 내용이나 부족한 부분이 있을까요? 검토 부탁드립니다.
> # 서드 파티 API를 통해 가져온 대용량 데이터 처리
> ### 대용량 데이터 사용처
> 1. 가격 추이 비교에 사용
>    
>    * 일단 MySQL에 저장하는 쿼리 ( Raw 데이터를 저장할 예정 )
>    * 저장 했던 데이터 중 날짜, 출발 공항, 도착 공항, 가격을 가지고 가격 추이 그래프 형성
> 2. 가격 변동 예측에 사용
>    
>    * 1번에서 뽑아온 데이터를 바탕으로 AI 혹은 알고리즘을 통해 가격 변동 예측 프로세스
> 
> ### 방법
> 1. 데이터 처리 속도가 느리다면 정형화된 데이터를 분산시켜서 관리하는건 어떤지?
>    
>    * 데이터를 통째로 보관해놓는 테이블과 가격추이나 변동예측에 사용하는 데이터를 분리시켜 보관 후 비교
> 2. Amadeus 에서 받아온 데이터는 데이터 형식이 MySQL로 처리하기에는 복잡한 부분이 있어 NoSQL을 사용해서 데이터 처리 속도 비교해보기
> 
> # SELECT 쿼리 수행 동작을 조사하고 부하테스트 및 성능 개선
> ## 선택 이유
> * SELECT 쿼리가 데이터베이스에서 어떻게 수행되는지를 깊이 이해하는 것은 프로젝트의 성능 최적화, 복잡한 데이터 구조의 효율적인 관리, 실시간 데이터 처리, 대규모 트래픽 처리에 직접적인 기여를 하며, 이러한 이유로 SELECT 쿼리 수행의 내부 과정을 깊이 있게 분석하는 것이 필수적 생각하여 선택했습니다.
> * ngrinder, jmeter를 사용하여 부하테스트를 진행하고, 성능개선을 해볼 생각입니다.
> 
> ## SELECT 쿼리의 기본 원리
> ### 데이터베이스 기본 구조
> ![image](https://private-user-images.githubusercontent.com/122856840/362563573-1e3b69c0-e072-4cd0-8577-8fe8a8baf023.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1MTk3MDQsIm5iZiI6MTcyNTUxOTQwNCwicGF0aCI6Ii8xMjI4NTY4NDAvMzYyNTYzNTczLTFlM2I2OWMwLWUwNzItNGNkMC04NTc3LThmZThhOGJhZjAyMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwNVQwNjU2NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMDRhYTZhN2M2NDJiZDFhYjk1YzI1MzgyNzRjMWZiNGZkNjU2ZjAzZmJmNzQ4OTE2NTRlNzEwMmNjNjMxZGUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.saOISgv7DfLaKYr558uef4D9zdV3zXlyVoUADQ3irXw)
> 
> ### MySQL 엔진 구조
> ![image](https://private-user-images.githubusercontent.com/122856840/362563876-0abdaa9d-1f6c-4638-896e-c0ac63bcb5d0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1MTk3MDQsIm5iZiI6MTcyNTUxOTQwNCwicGF0aCI6Ii8xMjI4NTY4NDAvMzYyNTYzODc2LTBhYmRhYTlkLTFmNmMtNDYzOC04OTZlLWMwYWM2M2JjYjVkMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwNVQwNjU2NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMGEyZmY5MjkxMjc3ZThiZTJhZWY1OTI4OTFhMWJkMmU5ZGFlMjI5ZjdhYTg2MTg4NTY1NWVlY2ExYzE2YWM2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.I2SefwBspflbFbLGU12xMp2q3jOeRd3u13njczh8qD8)
> 
> 1. 쿼리 파서
>    
>    1. 쿼리 문장을 MySQL이 인식할 수 있는 최소 단위인 토큰으로 분리 후 트리구조를 만듬
>    2. 전처리기에게 전달
>    3. SQL 문법 오류는 이 과정에서 발견되어 사용자에게 여러 메세지가 전달
> 2. 전처리기
>    
>    1. 전달받은 파서 트리를 기반으로 테이블 이름이 존재하는지, 컬럼 이름이 존재하는지, 접근 권한이 있는지 등을 확인
>    2. 이 부분에서 문제가 발견되면 오류처리
> 3. 옵티마이저
>    
>    1. 쿼리를 최적화해서 최종 실행계획 완성
>    2. `옵티마이저가 더 나은 선택을 할 수 있게 유도하는 방법` 조사
> 4. 쿼리 실행기
>    
>    1. 스토리지 엔진이 제공하는 핸들러 API를 가지고 옵티마이저가 계획해준 작업 목록을 수행
> 5. **핸들러 엔진**
>    
>    1. MySQL 서버에서 MySQL 엔진은 하나지만 스토리지 엔진은 여러개를 동시에 사용할 수 있다.
>    2. MySQL 엔진은 스토리지 엔진이 제공하는 API 를 이용하여 스토리지 엔진에게 CRUD 직업을 위임.
>    3. 이 때 스토리지 엔진이 제공하는 api를 핸들러 api 라고 한다.
>    4. 핸들러 규칙을 만족하도록만 구현하면 누구든지 스토리지 엔진을 만들고 MySQL 서버에 추가해서 사용할 수 있다.( **플러그인 모델 )**
>       
>       1. 스토리지 엔진 외에도 전문 검색 엔진을 위한 검색어 파서, 사용자 인증 등을 위한 플러그인 등 다양한 기능들이 플러그인으로 구현되어 제공
> 6. MySQL이 기본적으로 제공하는 스토리지 엔진
>    
>    1. InnoDB
> 
> ```
> // user 테이블에 대한 CRUD를 innoDB 스토리지 엔진이 처리하도록 한다.
> // 특정 테이블에 특정 스토리지 엔진을 지정가능
> create table user(
>   id bigint,
>   name varchar(50),
>   age int
> ) engine=innodb;
> ```
> 
> MySQL 스토리지 엔진
> #### 데이터베이스 인덱스 및 구조 이해
> ### SELECT 쿼리의 동작 방식
> #### 쿼리 플래닝과 실행 과정
> ### 옵티마이저의 역할
> #### 쿼리 최적화 및 실행 계획 수립

# 답변

- 테스트에 사용된 실제 쿼리와 테이블이 어떤 상태인지를 내용에 포함하는 것도 좋을 것 같습니다.
- 실제 테스트 데이터를 포함시켜주시고 그 과정 역시 상세하게 사용해주세요
